client/core/core.go

```diff
@@ -271,17 +274,71 @@ func (dc *dexConnection) compareServerMatches(matches map[order.OrderID]*serverM
 }
 
 // tickAsset checks open matches related to a specific asset for needed action.
-func (dc *dexConnection) tickAsset(assetID uint32) {
+func (dc *dexConnection) tickAsset(assetID uint32) (numUpdated int) {
 	dc.tradeMtx.RLock()
 	defer dc.tradeMtx.RUnlock()
 	for _, trade := range dc.trades {
 		if trade.Base() == assetID || trade.Quote() == assetID {
-			err := trade.tick()
+			n, err := trade.tick()
 			if err != nil {
 				log.Errorf("%s tick error: %v", dc.acct.url, err)
 			}
+			numUpdated += n
 		}
 	}
+	return
+}
+
+// market gets the *Market from the marketMap, or nil if mktID is unknown.
+func (dc *dexConnection) market(mktID string) *Market {
+	dc.marketMtx.RLock()
+	defer dc.marketMtx.RUnlock()
+	return dc.marketMap[mktID]
+}
+
+// setEpoch sets the epoch. If the passed epoch is greater than the highest
+// previously passed epoch, send an epoch notification to all subscribers.
+func (dc *dexConnection) setEpoch(mktID string, epochIdx uint64) bool {
+	dc.epochMtx.Lock()
+	defer dc.epochMtx.Unlock()
+	if epochIdx > dc.epoch[mktID] {
+		dc.epoch[mktID] = epochIdx
+		dc.notify(newEpochNotification(dc.acct.url, mktID, epochIdx))
+		var updateCount int
+		dc.tradeMtx.Lock()
+		for _, trade := range dc.trades {
+			if trade.processEpoch(epochIdx) {
+				updateCount++
+			}
+		}
+		dc.tradeMtx.Unlock()
+		if updateCount > 0 {
+			dc.refreshMarkets()
+			return true
+		}
+	}
+	return false
+}
+
+// marketEpochDuration gets the market's epoch duration. If the market is not
+// known, an error is logged and 0 is returned.
+func (dc *dexConnection) marketEpochDuration(mktID string) uint64 {
+	mkt := dc.market(mktID)
+	if mkt == nil {
+		log.Errorf("marketEpoch called for unknown market %s", mktID)
+		return 0
+	}
+	return mkt.EpochLen
+}
+
+// marketEpoch gets the best known epoch for the specified market and time
```

Comment:
"best known epoch" is a bit confusing.  Looks like the function returns the epoch index that includes the specified `stamp` time.